include/communicator_quda.h

```diff
@@ -0,0 +1,789 @@
+  if (err != CUDA_SUCCESS) {
+    const char *str;
+    cuGetErrorName(err, &str);
+    errorQuda("cuPointerGetAttributes returned error %s", str);
+  }
+
+  switch (memType) {
+  case CU_MEMORYTYPE_DEVICE: return false;
+  case CU_MEMORYTYPE_ARRAY: errorQuda("Using array memory for communications buffer is not supported");
+  case CU_MEMORYTYPE_UNIFIED: errorQuda("Using unified memory for communications buffer is not supported");
+  case CU_MEMORYTYPE_HOST:
+  default: // memory not allocated by CUDA allocaters will default to being host memory
+    return true;
+  }
+}
+
+inline void check_displacement(const int displacement[], int ndim)
+{
+  for (int i = 0; i < ndim; i++) {
+    if (abs(displacement[i]) > max_displacement) {
+      errorQuda("Requested displacement[%d] = %d is greater than maximum allowed", i, displacement[i]);
+    }
+  }
+}
+
+struct Communicator {
+
+  /**
+    The gpuid is static, and it's set when the default communicator is initialized.
+  */
+  static int gpuid;
+  static int comm_gpuid() { return gpuid; }
+
+  /**
+    Whether or not the MPI_COMM_HANDLE is created by user, in which case we should not free it.
+  */
+  bool user_set_comm_handle;
+
+  bool peer2peer_enabled[2][4] = {{false, false, false, false}, {false, false, false, false}};
+  bool peer2peer_init = false;
+
+  bool intranode_enabled[2][4] = {{false, false, false, false}, {false, false, false, false}};
+
+  /** this records whether there is any peer-2-peer capability
+      (regardless whether it is enabled or not) */
+  bool peer2peer_present = false;
+
+  /** by default enable both copy engines and load/store access */
+  int enable_peer_to_peer = 3;
+
+  /** sets whether we cap which peers can use peer-to-peer */
+  int enable_p2p_max_access_rank = std::numeric_limits<int>::max();
+
+  void comm_peer2peer_init(const char *hostname_recv_buf)
+  {
+    if (peer2peer_init) return;
+
+    // set gdr enablement
+    if (comm_gdr_enabled()) {
+      if (getVerbosity() > QUDA_SILENT && rank == 0) printf("Enabling GPU-Direct RDMA access\n");
+      comm_gdr_blacklist(); // set GDR blacklist
+      // by default, if GDR is enabled we disable non-p2p policies to
+      // prevent possible conflict between MPI and QUDA opening the same
+      // IPC memory handles when using CUDA-aware MPI
+      enable_peer_to_peer += 4;
+    } else {
+      if (getVerbosity() > QUDA_SILENT && rank == 0) printf("Disabling GPU-Direct RDMA access\n");
+    }
+
+    char *enable_peer_to_peer_env = getenv("QUDA_ENABLE_P2P");
+
+    // disable peer-to-peer comms in one direction if QUDA_ENABLE_P2P=-1
+    // and comm_dim(dim) == 2 (used for perf benchmarking)
+    bool disable_peer_to_peer_bidir = false;
+
+    if (enable_peer_to_peer_env) {
+      enable_peer_to_peer = atoi(enable_peer_to_peer_env);
+
+      switch (std::abs(enable_peer_to_peer)) {
+      case 0:
+        if (getVerbosity() > QUDA_SILENT && rank == 0) printf("Disabling peer-to-peer access\n");
```

Comment:
More a broad question first being prompted here: Is this `rank` the split-grid-local rank, or the "global" rank?

Reply:
This is the split-grid-local rank - `rank` is a field of the class.

Reply:
Seems like this'll become a mess of `printf`s really quickly when split grid is enabled...

Reply:
The tricky thing is this is part of the communicator constructor, so you can't call the `comm_rank_global()` or something ...